bundles/org.dataflowanalysis.analysis/src/org/dataflowanalysis/analysis/core/AbstractVertex.java

```diff
@@ -0,0 +1,224 @@
+     */
+    public List<DataFlowVariable> getAllOutgoingDataFlowVariables() {
+        return this.outgoingDataFlowVariables.orElseThrow(IllegalStateException::new);
+    }
+
+    /**
+     * Returns a list of all present node characteristics for the action sequence element
+     * @return List of present node characteristics
+     */
+    public List<CharacteristicValue> getAllNodeCharacteristics() {
+        return this.vertexCharacteristics.orElseThrow(IllegalStateException::new);
+    }
+
+    /**
+     * Returns the list of previous elements that precede the vertex
+     * @return Returns a list of all preceding vertices of the vertex.
+     */
+    public abstract List<? extends AbstractVertex<?>> getPreviousElements();
+
+    /**
+     * Returns whether the vertex is a source (e.g. does not have a source)
+     * @return Returns true, if the vertex is a source. Otherwise, the method returns false
+     */
+    public boolean isSource() {
+        return this.getPreviousElements().isEmpty();
+    }
+
+    /**
+     * Returns a list of characteristic literals that are set for a given characteristic type in the list of all node
+     * characteristics
+     * <p>
+     * See {@link AbstractVertex#getDataFlowCharacteristicNamesWithType(String)} for a similar method for data flow
+     * variables
+     * @param name Name of the characteristic type
+     * @return Returns a list of all characteristic literals matching the characteristic type
+     */
+    public List<String> getNodeCharacteristicNamesWithName(String name) {
+        return this.getAllNodeCharacteristics().stream().filter(cv -> cv.getTypeName().equals(name)).map(CharacteristicValue::getValueName)
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * Returns a list of characteristic literals that are set for a given characteristic type in the list of all node
+     * characteristics
+     * <p>
+     * See {@link AbstractVertex#getDataFlowCharacteristicIdsWithType(String)} for a similar method for data flow variables
+     * @param name Name of the characteristic type
+     * @return Returns a list of all characteristic literals matching the characteristic type
+     */
+    public List<String> getNodeCharacteristicIdsWithName(String name) {
+        return this.getAllNodeCharacteristics().stream().filter(cv -> cv.getTypeName().equals(name)).map(CharacteristicValue::getValueId)
+                .collect(Collectors.toList());
+    }
+
+    /**
+     * Returns a List of IDs of characteristics and data flow variables that are set for a given characteristic type in the
+     * list of all data flow variables
+     * <p>
+     * See {@link AbstractVertex#getNodeCharacteristicIdsWithName(String)} for a similar method for node characteristics
+     * @param type Name of the characteristic type
+     * @return Returns a list of all characteristic literals matching the characteristic type
```

Comment:
This returns a list of lists, copy paste error?

Reply:
Fixed in 8f74e1c836e0a214fc8c9e3f5ae0f2214ea2032d